doc/HCS.1
'\" t
.\"     Title: hcs
.\"    Author: Qball Cow qball@gmpclient.org
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 01/04/2014
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "HCS" "1" "01/04/2014" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
HCS \- Small command\-line tool to control a Voltcraft PPS\-11360 Programmable DC Power Supply\&.
.SH "SYNOPSYS"
.sp
hcs [\fIcommand\fR]
.SH "DESCRIPTION"
.sp
This manual page explains the \fBHCS\fR program\&. The \fBHCS\fR program provides basic control for the Programmable DC Power Supply\&.
.sp
The tool allows you to retrieve status from the power supply, set voltage, current and output state\&.
.SH "COMMANDS"
.sp
Multiple commands can be specified in one go\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBstatus\fR
Report status from the power supply\&. (Current voltage, current and active limiter)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBvoltage\fR
Read the output voltage\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBvoltage <value>\fR: Set the output voltage (in Volts)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBcurrent\fR
Read the output current limit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBcurrent <value>\fR: Set the output current limiter (in Amps)\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBoff\fR
Turn the output off\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBon\fR
Turn the output on\&.
.RE
.SH "EXAMPLES"
.sp
.if n \{\
.RS 4
.\}
.nf
hcs voltage 5 output 1\&.3 status
.fi
.if n \{\
.RE
.\}
.sp
Set the output voltage to 5Volts, the output current limit to 1\&.3Amps and report the current status\&.
.SH "ENVIRONMENT VARIABLES"
.sp
HCS obeys the following environment variables:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBHCS_DEVICE\fR
The device node pointing to the serial device of the programmable power supply\&.
.RE
.sp
\fIDefault:\fR
.sp
.if n \{\
.RS 4
.\}
.nf
/dev/ttyUSB0
.fi
.if n \{\
.RE
.\}
.SH "SUPPORTED DEVICES"
.sp
The following device(s) are known to work:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Voltcraft PPS\-11360
.RE
.SH "BUGS"
.sp
Please report bugs on the GITHUB issues page: https://github\&.com/DaveDavenport/hcs
.SH "LICENSE"
.sp
This program is licensed under GPL 2\&.0\&.
.SH "AUTHOR"
.PP
\fBQball Cow qball@gmpclient\&.org\fR
.RS 4
Author.
.RE


hcs.cc
/**
 *    This file is part of HCS.
 *    Written by Qball Cow <qball@gmpclient.org> 2013
 *
 *    HCS is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 2 of the License.
 *
 *    HCS is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with HCS.  If not, see <http://www.gnu.org/licenses/>.
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <sys/signal.h>
#include <sys/types.h>
#include <string.h>
#include <string>
#include <errno.h>

#define BAUDRATE B9600
#define MODEMDEVICE "/dev/ttyUSB0"

#define _POSIX_SOURCE 1

#define FALSE 0
#define TRUE 1


class HCS
{
    private:
        int fd = 0;
        struct termios oldtio;


        void send_cmd ( const char *command, const char *arg ) {
            // Check command.
            if ( command == nullptr ) return;

            // Write command to str.
            write( fd, command, strlen( command ) );

            if ( arg != nullptr ) {
                // Send argument.
                write( fd, arg, strlen( arg ) );
            }

            // End line
            write( fd, "\r" , 1 );
        }

        size_t read_cmd ( char *buffer, size_t max_length ) {
            size_t size = 0;

            while ( size < 3 ||
                    !(
                        buffer[size-3] == 'O' &&
                        buffer[size-2] == 'K' &&
                        buffer[size-1] == '\n'
                    )
                  ) {
                ssize_t v = read( fd, &buffer[size], max_length-size );
                buffer[size+1] = '\0';

                if ( buffer[size] == '\r' ) buffer[size] = '\n';

                if ( v > 0 ) {
                    size+=v;

                    if ( size+1 >= max_length ) {
                        return -1;
                    }
                } else {
                    printf( "%i\n", errno );
                    printf( "%s\n", strerror( errno ) );
                    return -1;
                }
            }

            return size;
        }


        void get_status ( void ) {
            char buffer[1024];
            // Send getd mesg.
            this->send_cmd( "GETD",NULL );

            // Get
            if ( this->read_cmd( buffer, 1024 ) > 0 ) {
                std::string b = buffer;
                double voltage = strtol( b.substr( 0,3 ).c_str(), 0, 10 )/10.0;
                double status = strtol( b.substr( 4,7 ).c_str(), 0, 10 )/1000.0;
                int limited = strtol( b.substr( 8,8 ).c_str(), 0, 10 );

                printf( "Volt: %.2f V\n",voltage );
                printf( "Curr: %.2f A\n",status );
                printf( "V*C:  %.2f VA\n",voltage*status );
                printf( "Lim:  %s\n",   ( limited == 0 )?"Voltage":"Current" );
            }
        }

        void set_on ( void ) {
            char buffer[1024];
            this->send_cmd( "SOUT", "0" );
            this->read_cmd( buffer, 1024 );
        }
        void set_off ( void ) {
            char buffer[1024];
            this->send_cmd( "SOUT", "1" );
            this->read_cmd( buffer, 1024 );
        }

        void set_volt ( float value ) {
            char buffer[1024];
            snprintf( buffer,1024,"%03d", ( int )( value*10 ) );
            this->send_cmd( "VOLT", buffer );
            this->read_cmd( buffer, 1024 );
        }
        void set_current ( float value ) {
            char buffer[1024];
            snprintf( buffer,1024,"%03d", ( int )( value*100 ) );
            this->send_cmd( "CURR", buffer );
            this->read_cmd( buffer, 1024 );
        }

        void get_voltage_current ( double &voltage, double &current ) {
            char buffer[1024];
            this->send_cmd( "GETS", buffer );
            voltage = current = -1.0;

            if ( this->read_cmd( buffer, 1024 ) > 5 ) {
                std::string b = buffer;
                voltage = strtol( b.substr( 0,3 ).c_str(), 0, 10 )/10.0;
                current= strtol( b.substr( 3,6 ).c_str(), 0, 10 )/100.0;
            }
        }

    public:
        HCS() {
            if ( getenv( "HCS_DEVICE" ) == nullptr ) {
                fd = open( MODEMDEVICE, O_RDWR | O_NOCTTY );
            } else {
                const char *path = getenv( "HCS_DEVICE" );
                fd = open( path, O_RDWR | O_NOCTTY );
            }

            if ( fd < 0 ) {
                printf( "failed: %s\n", strerror( errno ) );
                // TODO: Error checker.
                exit( -1 );
            }

            // save status port settings
            tcgetattr( fd,&oldtio );

            struct termios newtio;

            newtio.c_cflag = BAUDRATE | CS8;
            newtio.c_iflag = 0;
            newtio.c_oflag = 0;
            newtio.c_lflag = 0;       //ICANON;
            newtio.c_cc[VMIN]=1;
            newtio.c_cc[VTIME]=0;
            tcflush( fd, TCIFLUSH );
            tcsetattr( fd,TCSANOW,&newtio );
        }

        ~HCS() {
            if ( fd != 0 ) {
                // Restore
                tcsetattr( fd, TCSANOW, &oldtio );
                close( fd );
                fd=0;
            }
        }


        int run ( int argc, char **argv ) {
            for ( int i=0; i < argc; i++ ) {
                const char *command = argv[i];

                if ( strncmp( command, "status", 6 ) == 0 ) {
                    this->get_status();
                } else if ( strncmp( command, "on", 2 ) == 0 ) {
                    this->set_on();
                } else if ( strncmp( command, "off", 3 ) == 0 ) {
                    this->set_off();
                } else if ( strncmp( command, "voltage", 7 ) == 0 ) {
                    if ( argc > ( i+1 ) ) {
                        // write
                        const char *value = argv[++i];
                        float volt = strtof( value, nullptr );
                        this->set_volt( volt );
                    } else {
                        double voltage, current;
                        // read
                        this->get_voltage_current( voltage, current );
                        printf( "%.2f\n", voltage );
                    }
                } else if ( strncmp( command, "current", 7 ) == 0 ) {
                    if ( argc > ( i+1 ) ) {
                        // write
                        const char *value = argv[++i];
                        float current = strtof( value, nullptr );
                        this->set_current( current );
                    } else {
                        double voltage, current;
                        // read
                        this->get_voltage_current( voltage, current );
                        printf( "%.2f\n", current );
                    }
                }

            }

            return EXIT_FAILURE;
        }


};

int main ( int argc, char **argv )
{
    HCS hcs;
    return hcs.run( argc, argv );
}
